### ISSUE TITLE
Font fallbacks (css-esque)

### ISSUE THREAD

**Issue by elifitch** [mapbox/mapbox-gl-js#9931]
## Motivation

The failure state of font loading is confusing and frustrating. Fonts are loaded based on zoom level, and when a font fails to load (e.g. with a 404, common if a font is erroneously deleted or a style is shared to another Mapbox account) the map fails to render. Fine at one zoom level or map location, broken once a user navigates to a problematic spot. It's a problem that can hide until a map is published, receiving production traffic.

## Design Alternatives

We could interpret the text-font property as a font load stack as well as a glyph range stack. Currently, if my layer's `text-font` property looks like this `text-font: ["brandon grotesque", "impact", "open sans"]`, we will attempt to pull glyphs from `impact` if there are glyphs that aren't covered by `brandon grotesque`, and then finally we would fall back to glyphs from `open sans` if there are glyphs not covered by `impact`. We could also treat this as a font load stack, loading glyphs from `impact` if we fail to load `brandon grotesque`, and loading `open sans` if we fail to load `impact`. This change in how we interpret the spec may align better with existing user expectations since this is how font stacks work in CSS.

We could also serve a baseline font at the API level as a "last line of defense" but personally I'd prefer to go with the above option, and leave this firmly in implementors' hands.

**Comment by 1ec5** [comment]
The proposed design would match the [fallback heuristics on iOS and macOS](https://docs.mapbox.com/ios/api/maps/6.0.0/customizing-fonts.html#client-side-fonts) with locally-rendered glyphs.

**Comment by arindam1993** [comment]
The thing is though our fonts are server side rendered, and served by the Style API.
I can understand the need for fallbacks if the the rendering engine is looking for locally installed fonts and fails to find one, but if its server side shouldn't the fallback be on the server itself?

**Comment by elifitch** [comment]
## TLDR: 
Skip to Option 3. I think we should do option 3.

## Commentary
@arindam1993 if we went for a "universal fallback", as a last line of defense, then yes. Like if we decided that  we should never ever under any circumstances ever have renders fail due to a missing font, and we decide that \~comic sans\~ or something like that will always be our final fallback, no matter what's in the style, then *yes* I think we should handle this on the server.

I do think we'd need a client-side solution If we want a style's existing font stack to also be used as a request stack. e.g. for `text-font: ["brandon grotesque", "impact", "open sans"]` GL JS attempts to load `brandon grotesque`, falling back to `impact` in case of failure, which itself would fall back to `open sans` in case of failure, and then if `open sans` failed to load, we'd have our current behavior.

So I think this proposal has 3 permutations to it.

## Options

### Option 1: Fully client-side request stack

Here we'd implement a client-side only solution, using the existing font stack syntax to fall back to the next font, if the previous font in the stack failed to load (e.g. 404).

Pros:
* Fully user controlled
* Arguably aligns with user expectations around font-loading

Cons: 
* Still possible to hit an unrenderable map if no fonts in the stack are able to load
* Changes to GL JS means that users can't benefit without a version bump
* Potentially increased complexity to GL JS's font loading code

### Option 2: Server side universal fallback

Here we'd make no client-side changes, and we'd designate an agreeable font with great language coverage as a fallback if a font fails to load.

Pros:
* Users won't have to upgrade GL JS to benefit
* Great safety, will fully prevent unrenderable maps due to failed font loads

Cons:
* Users not able to select final fallback
* We have to make the design decision about what font to use

### Option 3: Client-side request stack combined with universal fallback

Here we'd implement a client-side only solution, using the existing font stack syntax to fall back to the next font, if the previous font in the stack failed to load (e.g. 404). We'd backfill that user control by *also* specifying a server-side universal fallback. 

Pros:
* Users won't have to upgrade GL JS to get basic benefits
* Great safety, will fully prevent unrenderable maps due to failed font loads
* Users still have great control over fallback structure
* Arguably aligns with user expectations around font-loading

Cons:
* Users not able to select final fallback, but mitigated by their ability to control fallbacks up to that point
* We have to make the design decision about what font to use

**Comment by elifitch** [comment]
@arindam1993 Any thoughts on the pros/cons outlined above?
